tsdb/engine/tsm1/compact_test.go

```diff
@@ -2306,66 +2323,196 @@ func TestDefaultPlanner_PlanOptimize_Level4(t *testing.T) {
+	require.Zero(t, cgLen, "compaction group length; PlanLevel(2)")
+	_, cgLen = cp.PlanLevel(3)
+	require.Zero(t, cgLen, "compaction group length; PlanLevel(3)")
 
-	for i, p := range expFiles1 {
-		if got, exp := tsm[0][i], p.Path; got != exp {
-			t.Fatalf("tsm file mismatch: got %v, exp %v", got, exp)
-		}
-	}
+	tsmP, pLenP := cp.Plan(time.Now().Add(-time.Second))
+	require.Zero(t, len(tsmP), "compaction group; Plan()")
+	require.Zero(t, pLenP, "compaction group length; Plan()")
+
+	tsm, pLen, _ := cp.PlanOptimize()
+	require.Equal(t, 1, len(tsm), "compaction group")
+	require.Equal(t, int64(len(tsm)), pLen, "compaction group length")
+	require.Equal(t, len(expFiles), len(tsm[0]), "expected TSM files")
 }
 
-func TestDefaultPlanner_PlanOptimize_Multiple(t *testing.T) {
+// This test is added to account for a single generation that has a group size
+// under 2 GB so it should be further compacted to a single file.
+func TestDefaultPlanner_PlanOptimize_SmallSingleGeneration(t *testing.T) {
+	// ~650 MB total group size
 	data := []tsm1.FileStat{
 		{
-			Path: "01-04.tsm1",
-			Size: 251 * 1024 * 1024,
+			Path: "01-05.tsm1",
+			Size: 300 * 1024 * 1024,
 		},
 		{
-			Path: "02-04.tsm1",
-			Size: 1 * 1024 * 1024,
+			Path: "01-06.tsm1",
+			Size: 200 * 1024 * 1024,
 		},
 		{
-			Path: "03-04.tsm1",
-			Size: 1 * 1024 * 1024,
+			Path: "01-07.tsm1",
+			Size: 100 * 1024 * 1024,
 		},
 		{
-			Path: "04-04.tsm1",
-			Size: 1 * 1024 * 1024,
+			Path: "01-08.tsm1",
+			Size: 50 * 1024 * 1024,
 		},
+	}
+
+	cp := tsm1.NewDefaultPlanner(
+		&fakeFileStore{
+			PathsFn: func() []tsm1.FileStat {
+				return data
+			},
+		}, tsdb.DefaultCompactFullWriteColdDuration,
+	)
+
+	expFiles := make([]tsm1.FileStat, 0)
+	for _, file := range data {
+		expFiles = append(expFiles, file)
+	}
+
+	_, cgLen := cp.PlanLevel(1)
+	require.Zero(t, cgLen, "compaction group length; PlanLevel(1)")
+	_, cgLen = cp.PlanLevel(2)
+	require.Zero(t, cgLen, "compaction group length; PlanLevel(2)")
+	_, cgLen = cp.PlanLevel(3)
+	require.Zero(t, cgLen, "compaction group length; PlanLevel(3)")
+
+	tsmP, pLenP := cp.Plan(time.Now().Add(-time.Second))
+	require.Zero(t, len(tsmP), "compaction group; Plan()")
+	require.Zero(t, pLenP, "compaction group length; Plan()")
+
+	tsm, pLen, gLen := cp.PlanOptimize()
+	require.Equal(t, 1, len(tsm), "compaction group")
+	require.Equal(t, int64(len(tsm)), pLen, "compaction group length")
+	require.Equal(t, int64(1), gLen, "generation of TSM files")
+	require.Equal(t, len(expFiles), len(tsm[0]), "expected TSM files")
+}
+
+// This test is added to account for a single generation that has a group size
+// under 2 GB and has less then level 4 files it should be further compacted to a single file.
+func TestDefaultPlanner_PlanOptimize_SmallSingleGenerationUnderLevel4(t *testing.T) {
+	// ~650 MB total group size
+	data := []tsm1.FileStat{
 		{
-			Path: "05-03.tsm1",
-			Size: 2 * 1024 * 1024 * 1024,
+			Path: "01-02.tsm1",
```

Comment:
As recommended above, reverse this ordering, as well, to make sure the file order (and size order) doesn't break anything.